core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -389,79 +389,84 @@ public CompletableFuture<Void> maybeInitialize() {
-                }
-                // Update the endOffset of the partition.
-                if (!cachedState.isEmpty()) {
-                    // If the cachedState is not empty, findNextFetchOffset flag is set to true so that any AVAILABLE records
-                    // in the cached state are not missed
-                    findNextFetchOffset.set(true);
-                    endOffset = cachedState.lastEntry().getValue().lastOffset();
-                    // In case the persister read state RPC result contains no AVAILABLE records, we can update cached state
-                    // and start/end offsets.
-                    maybeUpdateCachedStateAndOffsets();
-                } else {
-                    endOffset = startOffset;
+                    stateEpoch = partitionData.stateEpoch();
+
+                    List<PersisterStateBatch> stateBatches = partitionData.stateBatches();
+                    for (PersisterStateBatch stateBatch : stateBatches) {
+                        if (stateBatch.firstOffset() < startOffset) {
+                            log.error("Invalid state batch found for the share partition: {}-{}. The base offset: {}"
+                                    + " is less than the start offset: {}.", groupId, topicIdPartition,
+                                stateBatch.firstOffset(), startOffset);
+                            completeInitializationWithException(future, new IllegalStateException(String.format("Failed to initialize the share partition %s-%s", groupId, topicIdPartition)));
+                            return;
+                        }
+                        InFlightBatch inFlightBatch = new InFlightBatch(EMPTY_MEMBER_ID, stateBatch.firstOffset(),
+                            stateBatch.lastOffset(), RecordState.forId(stateBatch.deliveryState()), stateBatch.deliveryCount(), null);
+                        cachedState.put(stateBatch.firstOffset(), inFlightBatch);
+                    }
+                    // Update the endOffset of the partition.
+                    if (!cachedState.isEmpty()) {
+                        // If the cachedState is not empty, findNextFetchOffset flag is set to true so that any AVAILABLE records
+                        // in the cached state are not missed
+                        findNextFetchOffset.set(true);
+                        endOffset = cachedState.lastEntry().getValue().lastOffset();
+                        // In case the persister read state RPC result contains no AVAILABLE records, we can update cached state
+                        // and start/end offsets.
+                        maybeUpdateCachedStateAndOffsets();
+                    } else {
+                        endOffset = startOffset;
+                    }
+                    // Set the partition state to Active and complete the future.
+                    partitionState = SharePartitionState.ACTIVE;
+                    future.complete(null);
```

Comment:
It might be better not to complete the future while holding the write lock. The registered action for the `future` could potentially try to acquire another lock that is held by a different thread, which, in turn, could be waiting on the write lock - resulting in a deadlock.

In short, I prefer to narrow the scope of the write lock to avoid potential deadlock issues.

Reply:
Hi @chia7712 , should we not even complete futures when they are being completed with exceptions `completeInitializationWithException` within the write lock, for example - [here](https://github.com/apache/kafka/pull/17957/files#diff-83f8b0eae4289cefc39e38e4b85e64d49610ceb900eb456285f1ab8d240d9555R445) or IIUC, I should only move `future.complete(null)` within finally block after releasing the write lock and add an `if` statement that this should be performed only if future hasn't been completed yet.

Reply:
> should we not even complete futures when they are being completed with exceptions completeInitializationWithException within the write lock,

Yes, that should not hold the write lock too!

Reply:
I think that's going to cast the net a bit wider in this PR. I agree that it's a worthwhile thing to do.

Reply:
> The registered action for the future could potentially try to acquire another lock that is held by a different thread

Hmm, I don't see that should be a problem. There can anyways be multiple threads waiting on the same lock in other methods as well. Am I missing something here?

Reply:
> I don't see that should be a problem. There can anyways be multiple threads waiting on the same lock in other methods as well. Am I missing something here?

Yes, that’s a great question. One possible deadlock path from my perspective could occur as follows:

a_thread:

1. Holds the lock of a shared partition.
2. Attempts to complete a delayed share fetch.
3. Waits for the lock of the delayed share fetch.

b_thread:

1. Holds the lock of the delayed share fetch.
2. Attempts to access the shared partition.
3. Waits for the lock of the shared partition.

Based on these rules, the potential deadlock path could be as [this line](https://github.com/apache/kafka/blob/615f1a0bf999cd0b5ccf38c09e6afaba89538e9f/core/src/main/java/kafka/server/share/SharePartitionManager.java#L589). The PersisterStateManager thread acts as a_thread, while b_thread could be any thread attempting to complete a delayed share fetch while handling other types of requests.

At any rate, my point is that we should hold the lock only when absolutely necessary, even if we haven’t encountered an existing deadlock case yet. This minimizes the risk and improves overall robustness.

Reply:
> I'll make these change as part of issue https://issues.apache.org/jira/browse/KAFKA-18129 .

It seems KAFKA-18129 is intended to add a write lock to `maybeInitialize`. However, the additional discussion focuses on the appropriate scope of the write lock, particularly concerning the potential for deadlocks. If we can't reach a consensus on the lock scope, we could still add the write lock to `maybeInitialize` in this PR. It doesn't seem like a significant change to make at this point.

Reply:
Hi @chia7712 , maybe I am not understanding it correctly, so just want to clarify - In this example, https://github.com/apache/kafka/pull/17957#discussion_r1864856887 , a_thread holds the `lock.writeLock()` of a share partition, however b_thread will hold the fetch lock of the share partition. It won't hold the `lock.readLock()` of the share partition until the write lock by persister has been released. So how can there be a deadlock?
cc- @apoorvmittal10

Reply:
@adixitconfluent I haven't observed this scenario in the codebase, so the following discussion aims to prevent potential deadlocks in the future.

Let's assume we add an action that attempts to complete a delayed fetch to the future completed by `persister`. This means the `persister` thread holds the write lock of the shared partition and then requires the lock for the delayed fetch. Meanwhile, another thread holding the lock for the same delayed fetch needs to acquire the lock for the shared partition. This situation could lead to a deadlock based on my understanding.

Reply:
Thanks for explaining that. Yes, I agree this is a good future enhancement, **reiterating what you pointed out, https://issues.apache.org/jira/browse/KAFKA-18129 does not cause any deadlock in the current code**. I've opened a PR for the same https://github.com/apache/kafka/pull/18053